File: Ejercicios Resueltos/C++/ejercicio4.cpp
#include <iostream>
#include <string>

using namespace std;

struct Nodo {
    int dato;
    Nodo* izquierdo;
    Nodo* derecho;
};

Nodo* crearNodo(int valor) {
    Nodo* nuevo = new Nodo();
    nuevo->dato = valor;
    nuevo->izquierdo = NULL;
    nuevo->derecho = NULL;
    return nuevo;
}

Nodo* insertar(Nodo* raiz, int valor) {
    if (raiz == NULL) return crearNodo(valor);
    if (valor < raiz->dato) raiz->izquierdo = insertar(raiz->izquierdo, valor);
    else if (valor > raiz->dato) raiz->derecho = insertar(raiz->derecho, valor);
    return raiz;
}

void imprimir(Nodo* raiz, string prefijo, bool esUltimo) {
    if (raiz == NULL) return;
    cout << prefijo << (esUltimo ? "\\--> " : "|--> ") << raiz->dato << endl;
    string nuevoPrefijo = prefijo + (esUltimo ? "     " : "|    ");
    bool izquierdoEsUltimo = (raiz->derecho == NULL);
    imprimir(raiz->izquierdo, nuevoPrefijo, izquierdoEsUltimo);
    imprimir(raiz->derecho, nuevoPrefijo, true);
}

int contarHojas(Nodo* raiz) {
    if (raiz == NULL) return 0;
    if (raiz->izquierdo == NULL && raiz->derecho == NULL) return 1;
    return contarHojas(raiz->izquierdo) + contarHojas(raiz->derecho);
}

int main() {
    Nodo* raiz = NULL;
    raiz = insertar(raiz, 10);
    insertar(raiz, 5); insertar(raiz, 15);
    insertar(raiz, 2); insertar(raiz, 7);
    insertar(raiz, 12); insertar(raiz, 20);
    insertar(raiz, 1); insertar(raiz, 3);
    insertar(raiz, 18); insertar(raiz, 25);

    cout << "--- EJERCICIO 4 ---" << endl;
    cout << "Arbol Estructurado:" << endl;
    if (raiz != NULL) {
        cout << raiz->dato << endl;
        imprimir(raiz->izquierdo, "", raiz->derecho == NULL);
        imprimir(raiz->derecho, "", true);
    }
    
    cout << "\n-----------------------" << endl;
    cout << "La cantidad total de hojas (nodos finales) es: " << contarHojas(raiz) << endl;

    return 0;
}

File: Ejercicios Resueltos/Java/ejercicio4.java
class Nodo {
    int dato;
    Nodo izquierdo, derecho;
    public Nodo(int dato) {
        this.dato = dato;
        this.izquierdo = null;
        this.derecho = null;
    }
}

public class ejercicio4 {
    
    public static Nodo insertar(Nodo raiz, int valor) {
        if (raiz == null) return new Nodo(valor);
        if (valor < raiz.dato) raiz.izquierdo = insertar(raiz.izquierdo, valor);
        else if (valor > raiz.dato) raiz.derecho = insertar(raiz.derecho, valor);
        return raiz;
    }

    public static void imprimirConFlechas(Nodo raiz, String prefijo, boolean esUltimo) {
        if (raiz == null) return;
        System.out.println(prefijo + (esUltimo ? "\\--> " : "|--> ") + raiz.dato);
        String nuevoPrefijo = prefijo + (esUltimo ? "     " : "|    ");
        boolean izquierdoEsUltimo = (raiz.derecho == null);
        imprimirConFlechas(raiz.izquierdo, nuevoPrefijo, izquierdoEsUltimo);
        imprimirConFlechas(raiz.derecho, nuevoPrefijo, true);
    }

    public static int contarHojas(Nodo raiz) {
        if (raiz == null) return 0;
        if (raiz.izquierdo == null && raiz.derecho == null) return 1;
        return contarHojas(raiz.izquierdo) + contarHojas(raiz.derecho);
    }

    public static void main(String[] args) {
        Nodo raiz = null;
        raiz = insertar(raiz, 10);
        insertar(raiz, 5); insertar(raiz, 15);
        insertar(raiz, 2); insertar(raiz, 7);
        insertar(raiz, 12); insertar(raiz, 20);
        insertar(raiz, 1); insertar(raiz, 3);
        insertar(raiz, 18); insertar(raiz, 25);

        System.out.println("--- EJERCICIO 4 ---");
        System.out.println("Arbol Estructurado:");
        if (raiz != null) {
            System.out.println(raiz.dato);
            imprimirConFlechas(raiz.izquierdo, "", raiz.derecho == null);
            imprimirConFlechas(raiz.derecho, "", true);
        }
        
        System.out.println("\n-----------------------");
        System.out.println("La cantidad total de hojas es: " + contarHojas(raiz));
    }
}